internal/cluster/distribution/eks/eksadapter/node_pool_manager.go

```diff
@@ -125,42 +126,68 @@ func generateNodePoolStackName(clusterName string, poolName string) string {
+	c cluster.Cluster,
+	nodePoolStackIDs map[string]string,
+) (nodePools []eks.NodePool, err error) {
+	clusterClient, err := n.dynamicClientFactory.FromSecret(ctx, c.ConfigSecretID.String())
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed", "cluster", c)
 	}
 
 	manager := npls.NewManager(clusterClient, n.namespace)
 	labelSets, err := manager.GetAll(ctx)
 	if err != nil {
 		return nil, errors.WrapWithDetails(err, "retrieving node pool label sets failed",
-			"cluster", cluster,
+			"cluster", c,
 			"namespace", n.namespace,
 		)
 	}
 
-	awsClient, err := n.awsFactory.New(cluster.OrganizationID, cluster.SecretID.ResourceID, cluster.Location)
+	awsClient, err := n.awsFactory.New(c.OrganizationID, c.SecretID.ResourceID, c.Location)
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating aws factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating aws factory failed", "cluster", c)
 	}
 
 	cfClient := n.cloudFormationFactory.New(awsClient)
 	describeStacksInput := cloudformation.DescribeStacksInput{}
-	nodePools := make([]eks.NodePool, 0, len(nodePoolNames))
-	for _, nodePoolName := range nodePoolNames {
+	nodePools = make([]eks.NodePool, 0, len(nodePoolStackIDs))
+	for nodePoolName, nodePoolStackID := range nodePoolStackIDs {
 		nodePools = append(nodePools, eks.NodePool{
 			Name:   nodePoolName,
 			Labels: labelSets[nodePoolName],
 		})
 		nodePool := &nodePools[len(nodePools)-1]
 
-		stackName := generateNodePoolStackName(cluster.Name, nodePoolName)
-		describeStacksInput.StackName = &stackName
+		stackIdentifier := nodePoolStackID
+		if stackIdentifier == "" { // Note: CloudFormation stack creation not started yet.
+			stackIdentifier = generateNodePoolStackName(c.Name, nodePoolName)
+		}
+
+		describeStacksInput.StackName = aws.String(stackIdentifier)
 		stackDescriptions, err := cfClient.DescribeStacks(&describeStacksInput)
 		if err != nil {
-			nodePool.Status = eks.NodePoolStatusError
-			nodePool.StatusMessage = fmt.Sprintf("Retrieving node pool information failed: %s", err)
+			isCreating := nodePoolStackID == ""
+			hasError := c.Status == cluster.Warning || c.Status == cluster.Error
+
+			if isCreating && !hasError {
+				// Note: node pool is in the database already and the cluster is
+				// in a healthy state, thus CF stack creation has not been
+				// started yet due to the operation order.
+				nodePool.Status = eks.NodePoolStatusCreating
+				nodePool.StatusMessage = ""
+			} else if isCreating && hasError {
+				// Note: node pool creation failed before the CF stack creation
+				// would have been started thus no CF stack is available to
+				// describe.
+				nodePool.Status = eks.NodePoolStatusError
+				nodePool.StatusMessage = fmt.Sprintf("Creating node pool failed: %s", c.StatusMessage)
```

Comment:
**Note**: this is not correct in case the user issues an operation unrelated to the failed to create node pool and then lists the node pools.

So 
1. create invalid node pool
2. node pool create validation fails
3. list node pools (returns correct failure message)
4. create valid node pool || create differently invalid node pool || delete cluster || ...
5. list node pools (returns incorrect failure message as the last cluster status message is not the node pool failure message)

I'm going to think about how this could be done without stored node pool status messages, because I wanted to avoid that deliberately (probably that's the subconscious reason why I haven't caught this earlier).

I don't see any other solution right now and that (storing status message) feels a little bit overkill for the edge case of the edge case (I mean why would anyone leave a failed to create node pool as is), but for completeness this needs to be fixed.

Reply:
Yeah, the whole cluster status fallback is very fragile and not reliable due to possible other unrelated operations run after the describe failure node pool is introduced.

Right now I don't think we can avoid storing the node pool status and status message in the database as a fallback for cases when the describing the stack fails.

Reply:
[Fixed](https://github.com/banzaicloud/pipeline/compare/82c01a8999f6b30ef6da3428b2d2043d054ed781..f150e2ca28b67797e2d68b99c2593373ef1bbfc8).